pennylane_lightning_gpu/src/algorithms/AdjointDiffGPU.hpp

```diff
@@ -139,69 +139,587 @@ namespace Pennylane::Algorithms {
+     * object.
+     *
+     * @param Another instance of subclass of Observable<T> to compare
+     */
+    [[nodiscard]] virtual bool isEqual(const ObservableGPU<T> &other) const = 0;
+
+  protected:
+    ObservableGPU() = default;
+    ObservableGPU(const ObservableGPU &) = default;
+    ObservableGPU(ObservableGPU &&) noexcept = default;
+    ObservableGPU &operator=(const ObservableGPU &) = default;
+    ObservableGPU &operator=(ObservableGPU &&) noexcept = default;
+
   public:
+    virtual ~ObservableGPU() = default;
+
     /**
-     * @brief Variant type of stored parameter data.
+     * @brief Apply the observable to the given statevector in place.
      */
-    using param_var_t = std::variant<std::monostate, std::vector<T>,
-                                     std::vector<std::complex<T>>>;
+    virtual void applyInPlace(StateVectorCudaManaged<T> &sv) const = 0;
 
     /**
-     * @brief Copy constructor for an ObsDatum object, representing a given
-     * observable.
+     * @brief Get the name of the observable
+     */
+    [[nodiscard]] virtual auto getObsName() const -> std::string = 0;
+
+    /**
+     * @brief Get the wires the observable applies to.
+     */
+    [[nodiscard]] virtual auto getWires() const -> std::vector<size_t> = 0;
+
+    /**
+     * @brief Test whether this object is equal to another object
+     */
+    [[nodiscard]] bool operator==(const ObservableGPU<T> &other) const {
+        return typeid(*this) == typeid(other) && isEqual(other);
+    }
+
+    /**
+     * @brief Test whether this object is different from another object.
+     */
+    [[nodiscard]] bool operator!=(const ObservableGPU<T> &other) const {
+        return !(*this == other);
+    }
+};
+
+/**
+ * @brief Class models named observables (PauliX, PauliY, PauliZ, etc.)
+ *
+ * @tparam T Floating point type
+ */
+template <typename T> class NamedObsGPU final : public ObservableGPU<T> {
+  private:
+    std::string obs_name_;
+    std::vector<size_t> wires_;
+    std::vector<T> params_;
+
+    [[nodiscard]] bool isEqual(const ObservableGPU<T> &other) const override {
+        const auto &other_cast = static_cast<const NamedObsGPU<T> &>(other);
+
+        return (obs_name_ == other_cast.obs_name_) &&
+               (wires_ == other_cast.wires_) && (params_ == other_cast.params_);
+    }
+
+  public:
+    /**
+     * @brief Construct a NamedObsGPU object, representing a given observable.
      *
-     * @param obs_name Name of each operation of the observable. Tensor product
-     * observables have more than one operation.
-     * @param obs_params Parameters for a given observable operation ({} if
-     * optional).
-     * @param obs_wires Wires upon which to apply operation. Each observable
-     * operation will be a separate nested list.
+     * @param arg1 Name of the observable.
+     * @param arg2 Argument to construct wires.
+     * @param arg3 Argument to construct parameters
```

Comment:
```suggestion
     * @param obs_name Name of the observable.
     * @param wires Argument to construct wires.
     * @param params Argument to construct parameters
```

Reply:
Just a suggestion. I believe this can be improved, though.

Reply:
Thanks Amintor. Since we use this as part of the perfect forwarding with the same args in PennyLane Lightning, I think we keep this now and simplify it across all packages later. I do agree naming would be better here, but I'd like to mirror both packages and update everything we possible.

Reply:
Sure. I was wondering if we had a special reason for that.
Thanks for clarifying.